src/openfl/text/_internal/TextLayout.hx

```diff
@@ -153,6 +154,31 @@ class TextLayout
 			__hbBuffer.addUTF16(untyped __cpp__('(uintptr_t){0}', text.wc_str()), text.length, 0, -1);
 			#end
 
+			#if (!hl)
+			if (__script == TextScript.GUESS)
+			{
+				// For now I'm resettin like this, but I think we should keep the original user defined direction
+				// separated from the guessed direction, so we don't have to reset "manually"
+				// We should be able to have a Guess Script and defined direction
+				__hbBuffer.direction = INVALID;
+				__hbBuffer.language = "";
+				__hbBuffer.guessSegmentProperties();
+
+				if (__hbBuffer.script == INVALID)
+				{
+					direction = cast direction.toHBDirection();
+				}
+				else
+				{
+					direction = cast __hbBuffer.direction;
+				}
+
+				// This has to do with a weird bg in Guess mode where the first textLayout repeats itself
+				// instead of "ال صبا صباح"  it writes "ال ال صبا صباح "
+				if (positions.length > 0) return;
```

Comment:
Is it possible for `positions.length > 0` to be true here? `positions` is set to an empty array on line 103.

Reply:
Looks like the answer is yes, this is because of the `direction` setter which changes `dirty` to false. Therefore, everywhere else in this function where `positions` is accessed, its getter does another call to `__position`. This is a bit messy.

Reply:
If `__direction` is set directly instead of via the setter, this issue can be avoided, then maybe this check is no longer needed?

Reply:
For 'positions.length > 0`

I only put it in case of the "weird bug" I'm not sure I even had this problem in other non arabic text.

For _direction. I think it was necessary  Maybe it's for when you have only special characters like commas  where you can go both directions As the script would be invalid in this case,  the buffer direction would be ltr, when in reality it should be the direction of the text.

Reply:
> I only put it in case of the "weird bug" I'm not sure I even had this problem in other non arabic text.

I think this "weird bug" was happening because `direction =` accesses the `set_direction` setter, which has a side effect of setting `dirty = true`:
https://github.com/openfl/openfl/blob/88a12d2072fa4922a0cdd2ac161620e67b65936f/src/openfl/text/_internal/TextLayout.hx#L237

This means that all further accesses of the `positions` member variable (e.g. on [line 206](https://github.com/openfl/openfl/blob/88a12d2072fa4922a0cdd2ac161620e67b65936f/src/openfl/text/_internal/TextLayout.hx#L206)) in the `__position()` method, which pass through the `get_positions` getter, now call `__position()` again:

https://github.com/openfl/openfl/blob/88a12d2072fa4922a0cdd2ac161620e67b65936f/src/openfl/text/_internal/TextLayout.hx#L218-L222

So you end up with a `__position()` call in the middle of another `__position()` call, which is how it is possible for the `positions` array to no longer be empty even though it was emptied at the start of the function:
https://github.com/openfl/openfl/blob/88a12d2072fa4922a0cdd2ac161620e67b65936f/src/openfl/text/_internal/TextLayout.hx#L103


By using `__direction =` directly instead of `direction =` equals, this side effect can be avoided, and we don't end up with nested `__position()` calls:
https://github.com/tobil4sk/openfl/commit/e4bf39554c7671d5c7a550032bc3120c208b8818

Reply:
Shouldn't testing this be as easy as just adding two underlines before where we set the direction?

I might be able to test with Hebrew, just not in the next 12 hours (not at home rn)